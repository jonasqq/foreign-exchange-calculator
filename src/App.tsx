import * as React from 'react';
import { connect, Provider } from 'react-redux';
import Select from 'react-select';
import { TransitionGroup, CSSTransition } from 'react-transition-group';

import Page, { PageHeader, PageBody } from './components/Page';
import { Input, InputGroup } from './components/Field';
import CardContainer, { CurrencyCard } from './components/Card';
import Button from './components/Button';
import GlobalStyle from './components/GlobalStyle';

import { currencySelectOptions, altCurrencySelectOptions, getCurrencyLabel } from './helpers/currency';

import { fetchRate, addRateCard, deleteRateCard } from './redux/actions';
import { Rate, RateAction } from './redux/types';
import store from './redux/store';

const { useState, useEffect } = React;

interface MapState {
  rates: Rate[];
}
interface MapDispatch {
  fetchRate: (selectedCurrency: string, value: number) => RateAction;
  addRateCard: (currency: string) => RateAction;
  deleteRateCard: (cardIndex: number) => RateAction;
}

type Props = MapState & MapDispatch;

function AppPage({ rates, fetchRate, addRateCard, deleteRateCard }: Props): JSX.Element {
  const activeCurrencies = rates.map(rate => rate.currency);
  const altCurrencies = altCurrencySelectOptions.filter(option => !activeCurrencies.includes(option.value));
  const [form, changeForm] = useState({
    baseCurrency: currencySelectOptions[0],
    inputValue: 10,
    selectedNewCurrency: null,
  });
  const [showAddMoreBtn, toggleAddMoreBtn] = useState(true);
  const { baseCurrency, inputValue, selectedNewCurrency } = form;
  useEffect((): void => {
    fetchRate(baseCurrency.value, inputValue);
  }, [baseCurrency, inputValue, rates.length]);
  return (
    <Page>
      <GlobalStyle />
      <PageHeader>Foreign Exchange Rate</PageHeader>
      <PageBody style={{ marginBottom: 20 }}>
        <div style={{ fontStyle: 'italic', marginBottom: 5, fontSize: '0.8em' }}>
          {getCurrencyLabel(baseCurrency.value)}
        </div>
        <InputGroup style={{ marginBottom: 20 }}>
          <Select
            className="react-select"
            styles={{
              control: (style): React.CSSProperties => ({ ...style, height: 40, width: 100 }),
            }}
            onChange={(selected: { value: string; label: string }): void => {
              changeForm({ ...form, baseCurrency: selected });
            }}
            defaultValue={baseCurrency}
            classNamePrefix="select"
            isSearchable={true}
            name="currency"
            options={currencySelectOptions}
          />
          <Input
            style={{ width: 'calc(100% - 100px)', textAlign: 'right' }}
            type="number"
            name="currency-value"
            defaultValue={inputValue}
            onChange={(e): void => changeForm({ ...form, inputValue: parseInt(e.target.value) })}
          />
        </InputGroup>
        <CardContainer style={{ marginBottom: 10 }}>
          <TransitionGroup enter>
            {rates.map((rate: Rate, idx) => {
              return (
                rate.currency !== baseCurrency.value && (
                  <CSSTransition key={rate.currency} timeout={500} classNames="card">
                    <CurrencyCard
                      key={rate.currency}
                      baseCurrency={form.baseCurrency.value}
                      currency={rate.currency}
                      rate={rate.rate}
                      value={rate.rate * form.inputValue}
                      onDelete={(): void => {
                        deleteRateCard(idx);
                      }}
                    />
                  </CSSTransition>
                )
              );
            })}
          </TransitionGroup>
        </CardContainer>
        {showAddMoreBtn ? (
          <Button onClick={(): void => toggleAddMoreBtn(!showAddMoreBtn)}>(+) Add More Currencies</Button>
        ) : (
          <InputGroup style={{ textAlign: 'center' }}>
            <Select
              className="react-select"
              styles={{
                control: (style): React.CSSProperties => ({ ...style, height: 40, width: 100 }),
              }}
              onChange={(selected: { value: string; label: string }): void => {
                changeForm({ ...form, selectedNewCurrency: selected });
              }}
              value={selectedNewCurrency}
              classNamePrefix="select"
              isSearchable={true}
              name="add-currency-card"
              options={altCurrencies}
              placeholder="Select"
            />
            <Button
              onClick={(): void => {
                if (selectedNewCurrency) {
                  addRateCard(selectedNewCurrency.value);
                  changeForm({ ...form, selectedNewCurrency: null });
                }
              }}
            >
              Submit
            </Button>
          </InputGroup>
        )}
      </PageBody>
    </Page>
  );
}

const mapStateToProps = (state: { rates: Rate[] }): MapState => ({
  rates: state.rates,
});
const mapDispatchToProps = {
  fetchRate,
  addRateCard,
  deleteRateCard,
};
const App = connect(
  mapStateToProps,
  mapDispatchToProps,
)(AppPage);

export default function AppProvider(): JSX.Element {
  return (
    <Provider store={store}>
      <App />
    </Provider>
  );
}
