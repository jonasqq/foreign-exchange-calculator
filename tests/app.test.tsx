import * as React from 'react';
import * as renderer from 'react-test-renderer';
import App from '../src/App';

it('render correctly', () => {
  const app = renderer.create(<App />).toJSON();
  expect(app).toMatchSnapshot();
});
