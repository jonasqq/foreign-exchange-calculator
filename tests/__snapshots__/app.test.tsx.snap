// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render correctly 1`] = `
<div
  className="sc-bdVaJa gCggdw"
>
  <div
    className="sc-bwzfXH bbbxqf"
  >
    Foreign Exchange Rate
  </div>
  <div
    className="sc-htpNat dePyzG"
    style={
      Object {
        "marginBottom": 20,
      }
    }
  >
    <div
      style={
        Object {
          "fontSize": "0.8em",
          "fontStyle": "italic",
          "marginBottom": 5,
        }
      }
    >
      USD - United States Dollar
    </div>
    <div
      className="sc-ifAKCX laBfsI"
      style={
        Object {
          "marginBottom": 20,
        }
      }
    >
      <div
        className="react-select css-2b097c-container"
        onKeyDown={[Function]}
      >
        <div
          className="select__control css-f81n3h-control"
          onMouseDown={[Function]}
          onTouchEnd={[Function]}
        >
          <div
            className="select__value-container select__value-container--has-value css-1hwfws3"
          >
            <div
              className="select__single-value css-1uccc91-singleValue"
            >
              USD
            </div>
            <div
              className="css-b8ldur-Input"
            >
              <div
                className="select__input"
                style={
                  Object {
                    "display": "inline-block",
                  }
                }
              >
                <input
                  aria-autocomplete="list"
                  autoCapitalize="none"
                  autoComplete="off"
                  autoCorrect="off"
                  disabled={false}
                  id="react-select-2-input"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  spellCheck="false"
                  style={
                    Object {
                      "background": 0,
                      "border": 0,
                      "boxSizing": "content-box",
                      "color": "inherit",
                      "fontSize": "inherit",
                      "label": "input",
                      "opacity": 1,
                      "outline": 0,
                      "padding": 0,
                      "width": "1px",
                    }
                  }
                  tabIndex="0"
                  type="text"
                  value=""
                />
                <div
                  style={
                    Object {
                      "height": 0,
                      "left": 0,
                      "overflow": "scroll",
                      "position": "absolute",
                      "top": 0,
                      "visibility": "hidden",
                      "whiteSpace": "pre",
                    }
                  }
                >
                  
                </div>
              </div>
            </div>
          </div>
          <div
            className="select__indicators css-1hb7zxy-IndicatorsContainer"
          >
            <span
              className="select__indicator-separator css-1okebmr-indicatorSeparator"
            />
            <div
              aria-hidden="true"
              className="select__indicator select__dropdown-indicator css-tlfecz-indicatorContainer"
              onMouseDown={[Function]}
              onTouchEnd={[Function]}
            >
              <svg
                aria-hidden="true"
                className="css-6q0nyr-Svg"
                focusable="false"
                height={20}
                viewBox="0 0 20 20"
                width={20}
              >
                <path
                  d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                />
              </svg>
            </div>
          </div>
        </div>
        <input
          name="currency"
          type="hidden"
          value="USD"
        />
      </div>
      <input
        className="sc-bxivhb faemZt"
        defaultValue={10}
        name="currency-value"
        onChange={[Function]}
        style={
          Object {
            "textAlign": "right",
            "width": "calc(100% - 100px)",
          }
        }
        type="number"
      />
    </div>
    <div
      className="sc-gZMcBi jWFTjf"
      style={
        Object {
          "marginBottom": 10,
        }
      }
    >
      <div>
        <div
          className="sc-bZQynM kgkIPs"
        >
          <div
            className="sc-gzVnrw hpFAkG"
          >
            <span
              className="currency"
            >
              IDR
            </span>
            <span
              className="value"
            >
              0
            </span>
          </div>
          <div
            className="sc-htoDjs cnqSxK"
          >
            IDR - Indonesian Rupiah
          </div>
          <div
            className="sc-dnqmqq fOeLjC"
          >
            1 
            USD
             = 
            IDR
             
            0
          </div>
          <button
            className="sc-EHOje sc-iwsKbI XxEAE"
            onClick={[Function]}
          >
            (-)
          </button>
        </div>
        <div
          className="sc-bZQynM kgkIPs"
        >
          <div
            className="sc-gzVnrw hpFAkG"
          >
            <span
              className="currency"
            >
              EUR
            </span>
            <span
              className="value"
            >
              0
            </span>
          </div>
          <div
            className="sc-htoDjs cnqSxK"
          >
            EUR - Euro
          </div>
          <div
            className="sc-dnqmqq fOeLjC"
          >
            1 
            USD
             = 
            EUR
             
            0
          </div>
          <button
            className="sc-EHOje sc-iwsKbI XxEAE"
            onClick={[Function]}
          >
            (-)
          </button>
        </div>
        <div
          className="sc-bZQynM kgkIPs"
        >
          <div
            className="sc-gzVnrw hpFAkG"
          >
            <span
              className="currency"
            >
              GBP
            </span>
            <span
              className="value"
            >
              0
            </span>
          </div>
          <div
            className="sc-htoDjs cnqSxK"
          >
            GBP - British Pound
          </div>
          <div
            className="sc-dnqmqq fOeLjC"
          >
            1 
            USD
             = 
            GBP
             
            0
          </div>
          <button
            className="sc-EHOje sc-iwsKbI XxEAE"
            onClick={[Function]}
          >
            (-)
          </button>
        </div>
        <div
          className="sc-bZQynM kgkIPs"
        >
          <div
            className="sc-gzVnrw hpFAkG"
          >
            <span
              className="currency"
            >
              SGD
            </span>
            <span
              className="value"
            >
              0
            </span>
          </div>
          <div
            className="sc-htoDjs cnqSxK"
          >
            SGD - Singapore Dollar
          </div>
          <div
            className="sc-dnqmqq fOeLjC"
          >
            1 
            USD
             = 
            SGD
             
            0
          </div>
          <button
            className="sc-EHOje sc-iwsKbI XxEAE"
            onClick={[Function]}
          >
            (-)
          </button>
        </div>
      </div>
    </div>
    <button
      className="sc-EHOje gejnCW"
      onClick={[Function]}
    >
      (+) Add More Currencies
    </button>
  </div>
</div>
`;
